Requeue

mudler opened the pull request:
This introduces a controller which requeues when there are changes to `ManagedOSVersionChannel`, thus reverts back to default sync_interval in our tests(`60m`), as now that is not needed anymore.

It also ties the necessary pieces to report back at kubernetes, and the reconciler now reports back the status as well with events.

~~Still missing specific tests for it~~

Fixes: #23

mudler committed:
Add reconciler

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Build general event interface from raw k8s into client

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Record invalid specs back to the VersionChannel

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
go gen

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Set appropriate rules for broadcasting events

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Add test for event broadcasting

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Refactor out requeuer to not be blocking

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Collect errors when syncing

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Refactor out recorder boilerplate

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

Itxaka commented on pkg/controllers/managedosversionchannel/managedosversionchannel.go:
```diff
@@ -0,0 +1,66 @@
+	if moc.Spec.Type == "" {
+		copy := moc.DeepCopy()
+		copy.Status.Status = "error"
+		recorder.Event(moc, corev1.EventTypeWarning, "error", "No ManagedOSVersionChannel type defined")
```
so when recording an event we need to pass a deepcopy of the object? does it accept any type of object or its an interface type?

mudler replied:
It accepts a `runtime.Object` , so everything that satisfies that interface is good. Passing a DeepCopy is just a common pattern to avoid overriding the original object

mudler committed:
Do allocate the event recorder once in the syncer

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

Itxaka commented on pkg/services/syncer/syncer.go:
```diff
@@ -90,6 +94,7 @@ func syncNamespace(config config.Config, namespace string) error {
 
 		vers, err := s.Sync(config, cc)
 		if err != nil {
+			config.Recorder.Event(&cc, corev1.EventTypeWarning, "sync", err.Error())
```
so linking with the comment above, why in here we are not passing a copy but a pointer? will the Event modify the original object then?

mudler replied:
no, the difference from the above case is that we were modifying the status before updating it, that's why we took the DeepCopy, it should be safe here

mudler replied:
to wrap up (a generic rule of thumb) you deepcopy when you are about to change something that needs to be posted back at Kube: 
```
copy := moc.DeepCopy()
copy.Status.Status = "..."
//// then do something with copy, as posting an update and such
```

of course there might be other cases too, but this is generally a good way to go. You see usually also deepcopy intentionally also at beginning to avoid forgetting it by mistake

Itxaka replied:
ok, got it, thanks for the explanation!

mudler committed:
Respect upgradeContainerSpec from ManagedOSVersion

Before we were completely ignoring the upgradeContainerSpec field.

Now we correctly populate it while creating new resources and while
refering to the ManagedOSVersion from ManagedOS

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Adapt tests, add test cases

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Correctly annotate env vars from metadata

We were completely ignoring them before

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Test env metadata injection

Also refactor to quiet the linter

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler commented:
couple of unrelated fixes in here that we missed (tests added):
- Metadata env injection was not functional at all
- Upgrade container spec injection was not functional at all
- Refactored some code parts
- Added e2e tests for new parts and things that I've been fixing up, there are still some areas uncovered, but it is out of the purpose of this PR

mudler committed:
Do not make kube calls blocking

associate a context to them

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler commented:
@Itxaka @davidcassany @mjura Seems all good here from local testing too, looks good for review

Itxaka reviewed (approved):
codewise looks good but workflow wise Ima  bit lost in all this stuff so I cannot comment on that.

mjura reviewed (approved):
LGTM

mudler committed:
Update README

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler commented:
> codewise looks good but workflow wise Ima bit lost in all this stuff so I cannot comment on that.

I'm preparing some dev docs around it on how this should be used in the context of https://github.com/rancher-sandbox/os2/issues/87, as this is not really meant to be exposed to the user it doesn't feel to go into front README docs.
